linkerd/app/outbound/src/orig_proto_upgrade.rs

```diff
@@ -1,68 +1,72 @@
-use super::Endpoint;
+use crate::endpoint::Endpoint;
 use crate::proxy::http::{orig_proto, settings::Settings};
 use crate::svc;
 use futures::{try_ready, Future, Poll};
-use http;
-use std::marker::PhantomData;
 use tracing::trace;
 
-#[derive(Debug)]
-pub struct Layer<A, B>(PhantomData<fn(A) -> B>);
+#[derive(Clone, Debug, Default)]
+pub struct OrigProtoUpgradeLayer(());
 
-#[derive(Debug)]
-pub struct MakeSvc<M, A, B> {
+#[derive(Clone, Debug)]
+pub struct OrigProtoUpgrade<M> {
     inner: M,
-    _marker: PhantomData<fn(A) -> B>,
 }
 
-pub struct MakeFuture<F, A, B> {
+pub struct UpgradeFuture<F> {
     can_upgrade: bool,
     inner: F,
-    _marker: PhantomData<fn(A) -> B>,
+    was_absolute: bool,
 }
 
-pub fn layer<A, B>() -> Layer<A, B> {
-    Layer(PhantomData)
-}
-
-impl<A, B> Clone for Layer<A, B> {
-    fn clone(&self) -> Self {
-        Layer(PhantomData)
+impl OrigProtoUpgradeLayer {
+    pub fn new() -> Self {
+        Self::default()
     }
 }
 
-impl<M, A, B> svc::Layer<M> for Layer<A, B>
-where
-    M: svc::MakeService<Endpoint, http::Request<A>, Response = http::Response<B>>,
-{
-    type Service = MakeSvc<M, A, B>;
+impl<M> svc::Layer<M> for OrigProtoUpgradeLayer {
+    type Service = OrigProtoUpgrade<M>;
 
     fn layer(&self, inner: M) -> Self::Service {
-        MakeSvc {
-            inner,
-            _marker: PhantomData,
-        }
+        Self::Service { inner }
     }
 }
 
-// === impl MakeSvc ===
+// === impl OrigProtoUpgrade ===
 
-impl<M: Clone, A, B> Clone for MakeSvc<M, A, B> {
-    fn clone(&self) -> Self {
-        MakeSvc {
-            inner: self.inner.clone(),
-            _marker: PhantomData,
+impl<N> svc::NewService<Endpoint> for OrigProtoUpgrade<N>
+where
+    N: svc::NewService<Endpoint>,
+{
+    type Service = svc::Either<orig_proto::Upgrade<N::Service>, N::Service>;
+
+    fn new_service(&self, mut endpoint: Endpoint) -> Self::Service {
+        if !endpoint.can_use_orig_proto() {
+            trace!("Endpoint does not support transparent HTTP/2 upgrades");
+            return svc::Either::B(self.inner.new_service(endpoint));
         }
+
+        let was_absolute = endpoint.http_settings.was_absolute_form();
+        trace!(
+            header = %orig_proto::L5D_ORIG_PROTO,
+            %was_absolute,
```

Comment:
nit: `was_absolute` is a `bool`, right? `bool` implements `tracing::field::Value`, so we don't need to use `fmt::Display`:
```suggestion
            was_absolute,
```